README.md

# dyntaper  -  Using a dynamic tree stem profile model

<!-- badges: start -->
<!-- badges: end -->

An *R* package that implements the model of García (2015).

The main functions are *taper()*, which calculates diameters at specified height levels up the tree, its inverse *hlevel()*, which gives the height level for a given diameter, and *volume()*, which computes the volume between two height levels.

For a tutorial, see the [vignette](https://github.com/ogarciav/dyntaper/blob/master/doc/dyntaper.pdf).

## Installation

You can install dyntaper like so:

``` r
install.packages("devtools")
devtools::install_github("ogarciav/dyntaper")
```

## Example

This plots a diameter profile for a tree with a total height of 32 m, and a dbh of 24 cm (breast height 1.3 m):

``` r
library(dyntaper)
prms <- c(2.569, 0, 1.042, 0.3012, -1)  # parameters
curve(
  taper(h=x, H=32, D=24, b=prms, bh=1.3),
from=0, to=32)
```
## References

García, O. (2015) "Dynamic modelling of tree form". *MCFNS 7*, 9–15.
  (http://mcfns.net/index.php/Journal/article/view/MCFNS7.1_2).
  
Koirala, A., Montes, C. R., Bullock, B. P. and Wagle, B. H. (2021) "Developing
  taper equations for planted teak (*Tectona grandis* L. f.) trees of
  central lowland Nepal". *Trees, Forests and People 5*.
  (https://doi.org/10.1016/j.tfp.2021.100103).


inst/NEWS
Changes in version 1.1  (2022-07-30)
  - Version submitted to CRAN.
  - Minor documentation tweaks.
Changes in version 1.0  (2022-03-13)
  - First release.


R/decay.R
#' Decay function delta
#'
#' Calculates \eqn{(1 - p*x)_+^{1/p}}, or its limit \eqn{exp(-x)}
#'   when \eqn{p} tends to 0.
#'
#' @param x   Input value(s), possibly a vector.
#' @param p   Parameter.
#'
#' @return    Decay function value(s).
#' @export
#'
#' @details   Perhaps overkill, but uses \code{log1p()} function for better
#'              accuracy than the more obvious formula.
#' @examples
#'    decay(2, 0) == exp(-2)
#'    decay(1.5, 0.5)
#'    decay(2.5, 0.5)
#'    decay(2.5, -0.5)
#'    for(p in seq(1, -1, -0.5)) curve(decay(x, p), 0, 3, add=(p != 1))
#'
decay <- function(x, p){
  stopifnot(length(p) == 1) # sanity check
  if(abs(p) < 1e-300) exp(-x)
  else{
    suppressWarnings( # trap NaN's message
      v <- exp(log1p(-p * x) / p) # "same" as (1 - p*x)^(1/p)
    )
    v[is.nan(v)] <- 0  # when 1 - p*x <= 0
    return(v)
  }
}

#' Integral of decay function.
#'
#' @param x   Input value(s), possibly a vector.
#' @param p   Parameter.
#'
#' @return    Integral of the decay function between 0 and x.
#' @export
#'
#' @examples
#'    Id(2, 0)
#'    Id(1.5,0.5)
#'    Id(2.5, 0.5)
#'    Id(2.5, -0.5)
#'    for(p in seq(1, -1, -0.5)) curve(Id(x, p), 0, 3, add=(p != 1))
#'
Id <- function(x, p){
  if(abs(p + 1) < 1e-8) log1p(pmax(x, -1))
  else (1 - decay((p + 1) * x, p / (p + 1))) / (p + 1)
}

#' Double integral of decay function.
#'
#' @param x   Input value(s), possibly a vector.
#' @param p   Parameter.
#'
#' @return    Iterated integral of the decay function between 0 and x,
#'                that is, the integral of Id(x, p).
#' @export
#' @examples
#'    Idd(2, 0)
#'    Idd(1.5,0.5)
#'    Idd(2.5, -1)
#'    Idd(2.5, -0.5)
#'    for(p in seq(1, -1, -0.5)) curve(Idd(x, p), 0, 3, add=(p != 1))
#'
Idd <- function(x, p){
  if(abs(p + 1) < 1e-8){  # p = -1 (approx)
       suppressWarnings(v <- (x + 1) * log1p(x) - x)  # ignore NaN's
       v[is.nan(v)] <- Inf  # x <= -1
       v[x == -1] <- 1  # just in case
  }
  else v <- (x - Id((p + 1) * x, p / (p + 1)) / (p + 1)) / (p + 1)
  return(v)
}


R/taper.R
#' Tree taper (or profile) equation.
#'
#' Returns the diameter or cross-sectional area at one or more hight levels.
#' @note Diameters or areas are either all outside bark, or all inside bark.
#'   A dbh outside bark can be used with an inside-bark taper equation through
#'   a substitution \code{D -> k * D}, where \code{k} is an outside to inside
#'    bark conversion factor.
#'
#' @param h     Height level(s), possibly a vector.
#' @param H     Tree total height.
#' @param D     Tree diameter at breast height (dbh).
#' @param b     Vector with the 5 parameters.
#' @param bh    Breast height. Typically 1.2, 1.3 or 1.4 m, or 4.5 ft.
#' @param area  If TRUE, returns cross-sectional areas, otherwise returns
#'                diameters. Default is FALSE.
#'
#' @return  Diameter(s) at level(s) '\code{h}' if '\code{area}' is FALSE,
#'            otherwise cross-sectional area(s).
#' @export
#'
#' @examples
#'   curve(taper(x, 32, 0.956*24, c(2.569, 0, 1.042, 0.3012, -1), 1.3), 0, 32)
#'
taper <- function(h, H, D, b, bh, area = FALSE){
  r <- tbase(h, H, b) / tbase(bh, H, b) # relative area
  r[h < 0 | h > H] <- 0  # outside the stem
  if(area) pi * (D/2)^2 * r  # area at level h
  else D * sqrt(r)  # diameter at level h
}

#' Unscaled base taper curve of cross-sectional area vs height level.
#'
#' @param h     Height level(s), possibly a vector.
#' @param H     Tree total height.
#' @param b     Vector with the 5 parameters.
#'
#' @return  Value(s) proportional to the area(s) at the level(s) \code{h}.
#' @export
#'
#' @examples
#'     tbase(16, 32, c(2.569, 0, 1.042, 0.3012, -1))
#'
tbase <- function(h, H, b){
  H - h - b[1] * Id((H - h) / b[1], b[2]) +
    b[3] * (H - h) * decay(h / b[4], b[5])
}




R/hlevel.R
#' Height level for given diameter or area.
#'
#' Find the height level at which the tree reaches a given diameter (if
#'     \code{area == FALSE}, default) or a given cross-sectional area (if
#'     \code{area == TRUE}).
#'     That is, the inverse of \code{taper()}, except that a vector argument
#'     is not allowed.
#'
#' @param ds    The given diameter or area.
#' @param H     Tree total height.
#' @param D     Tree diameter at breast height.
#' @param b     Vector with the 5 parameters.
#' @param bh    Breast height. Typically 1.2, 1.3 or 1.4 m, or 4.5 ft.
#' @param area  If TRUE, '\code{ds}' is a cross-sectional area, otherwise a
#'                diameter. Default is FALSE.
#'
#' @return      Height level corresponding to '\code{ds}' if any, otherwise \code{NA}.
#' @export
#' @importFrom stats uniroot
#'
#' @examples
#'    b <- c(2.569, 0, 1.042, 0.3012, -1)  # parameters
#'    Dib <- 0.956 * 24  # convert dbh outside bark to inside bark 
#'    hlevel(15, 32, Dib, b, 1.3)  # height where the diameter is 15
#'    hlevel(24, 32, Dib, b, 1.3)  # breast height
#'    hlevel(35, 32, Dib, b, 1.3)  # height where the diameter is 35
#'
hlevel <- function(ds, H, D, b, bh, area = FALSE){
  f <- function(h, ds, H, D, b, bh, area){ # find zero of this over h
    ds - taper(h, H, D, b, bh, area)
  }
  stopifnot(length(ds) == 1) # vector not allowed
  if(sign(f(0, ds, H, D, b, bh, area)) == sign(f(H, ds, H, D, b, bh, area)))
    return(NA) # or NaN?
  uniroot(f, c(0, H), ds, H, D, b, bh, area, f.upper = ds)$root
}


R/volume.R
#' Volume between two height levels.
#'
#' @param h1,h2   The two height levels.
#' @param H       Tree total height.
#' @param D       Tree diameter at breast height.
#' @param b       Vector with the 5 parameters.
#' @param bh      Breast height. Typically 1.2, 1.3 or 1.4 m, or 4.5 ft.
#' @param rhd     Ratio between the height and diameter units. E.g., 100
#'                    for m and cm, or 12 for feet and inches.
#'
#' @return    Volume between \code{h1} and \code{h2}.
#' @export
#'
#' @examples
#'    b <- c(2.569, 0, 1.042, 0.3012, -1)  # parameters
#'    Dib <- 0.956 * 24  # convert dbh outside bark to inside bark 
#'    volume(0, 32, 32, Dib, b, 1.3, 100)  # total volume
#'    h15 <- hlevel(15, 32, Dib, b, 1.3)  # height for diameter 15
#'    volume(0.3, h15, 32, Dib, b, 1.3, 100)  # volume between stump and h15
#'
volume <- function(h1, h2, H, D, b, bh, rhd){
  Is <- function(h, H, b){ # indefinite integral of tbase()
    H * h - h^2 / 2 + b[1]^2 * Idd((H - h) / b[1], b[2]) + b[3] * b[4] *
      ((H - h) * Id(h / b[4], b[5]) + b[4] * Idd(h / b[4], b[5]))
  }
  h1 <- min(max(h1, 0), H) # force to within the stem, if necessary
  h2 <- min(max(h2, 0), H)
  (1 / rhd^2) * pi * (D / 2)^2 * abs(Is(h2, H, b) - Is(h1, H, b)) /
    tbase(bh, H, b)
}


R/dyntaper-package.R
#' @details
#' The main functions relevant to the user are
#' \itemize{
#'   \item \code{\link{taper}}: Calculates diameters or cross-sectional areas
#'     for given height levels.
#'   \item \code{\link{hlevel}}: Calculates the height level for a given
#'     diameter or cross-sectional area.
#'   \item \code{\link{volume}}: Calculates the volume between two given height
#'     levels.
#' }
#' The package includes a small example dataset, \code{\link{brink}}.
#'
#' For a tutorial, see the vignette: \code{vignette("dyntaper", package="dyntaper")}.
#'
#' @examples
#'    dfir <- c(2.569, 0, 1.042, 0.3012, -1)  # params. for D.fir in B.C.
#'    height <- 32; dbh <- 24;  k <- 0.956  # k is a bark conversion factor
#'    # Plot a taper curve (inside-bark diameter over height levels):
#'    curve(
#'      taper(h, height, k*dbh, dfir, 1.3),  # 1.3 m is the breast height
#'    from = 0, to = 32, xname = "h")
#'    # Height level for a diameter of 10 cm (i.b.):
#'    (h10 <- hlevel(10, height, k*dbh, dfir, 1.3))
#'    # Volume from a 30 cm stump to a 10 cm diameter limit:
#'    volume(0.3, h10, height, k*dbh, dfir, 1.3, 100)
#'
#' @references
#' García, O. (2015) "Dynamic modelling of tree form". \emph{Mathematical
#'   and Computational Forestry & Natural-Resource Sciences (MCFNS) 7}(1), 9–15.
#'   (\url{http://mcfns.net/index.php/Journal/article/view/MCFNS7.1_2}).
#'
#' Koirala, A., Montes, C. R., Bullock, B. P. and Wagle, B. H. (2021) "Developing
#'   taper equations for planted teak (\emph{Tectona grandis} L. f.) trees of
#'   central lowland Nepal". \emph{Trees, Forests and People 5}.
#'   (\doi{10.1016/j.tfp.2021.100103}).
#'
#' @keywords internal
"_PACKAGE"

## usethis namespace: start
## usethis namespace: end
NULL
